### PR TITLE
Support for Slurm batchSystem (resolves #682)

### LINKED ISSUE #682: SLURM integration
Hi,

I was wondering if there is any plan to adapt a SLURM batch system into toil. I have noticed that you evaluated this a while ago in #18 and #19 , and he conclusion was that SLURM does not meet the use case for Toil, that works with scalable systems like amazon EC2.

I've seen though that you guys provide an abstraction of batch system class [here](https://github.com/BD2KGenomics/toil/blob/master/src/toil/batchSystems/abstractBatchSystem.py). And there is for example [an implementation for LSF](https://github.com/BD2KGenomics/toil/blob/master/src/toil/batchSystems/lsf.py).

Unfortunately the documentation on how to implement this class is [uncomplete](http://toil.readthedocs.org/en/latest/batchSystem.html), but having those examples should be possible to do it for SLURM.

Do you guys think its possible and see a fit for this?

Thanks a lot!

### PR DESCRIPTION
**Pull request by dleehr**
Implements support for slurm scheduler

- SlurmBatchSystem class is modeled after GridengineBatchSystem
- Maps AbstractBatchSystem methods to sbatch, squeue, scancel, sacct, sinfo subprocess calls
- Includes SlurmBatchSystemTest for testing system

### CODE REVIEW

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
@dleehr I invited you to our organization. Once you accept the invitation, you should be able to see the Jenkins build results. I'm going to trigger another build, since the last build left the slave in a weird state.

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
@dleehr, previous builds had crashed the slave but the most recent one didn't. I'll try a few more times. In either case, these Slurm tests fail. Let me know if you don't see the 

Oh, before I forget you don't need to open a new PR after rebasing. By default, PR branches can be force-pushed. Only, PRs that are labelled `multi-author` are prohibited from being force pushed to.

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
@dleehr, let me know if you can't access see the build results on our Jenkins server

http://jenkins.cgcloud.info/job/toil-pull-requests/912/

**Comment by dleehr** [comment]
I can't tell why this fails without some additional output. The tests pass on my cluster (running slurm 14.11.9):
```
$ python setup.py test --pytest-args "-vv src/toil/test/batchSystems/batchSystemTest.py::SlurmBatchSystemTest"
running test
running egg_info
writing requirements to src/toil.egg-info/requires.txt
writing src/toil.egg-info/PKG-INFO
writing top-level names to src/toil.egg-info/top_level.txt
writing dependency_links to src/toil.egg-info/dependency_links.txt
writing entry points to src/toil.egg-info/entry_points.txt
reading manifest file 'src/toil.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'src/toil.egg-info/SOURCES.txt'
running build_ext
============================================================================================================== test session starts ==============================================================================================================
platform linux2 -- Python 2.7.3, pytest-2.8.3, py-1.4.31, pluggy-0.3.1 -- /home/dcl9/env-toil/bin/python
cachedir: .cache
rootdir: /home/dcl9/toil, inifile: setup.cfg
collected 5 items 

src/toil/test/batchSystems/batchSystemTest.py::SlurmBatchSystemTest::testAvailableCores PASSED
src/toil/test/batchSystems/batchSystemTest.py::SlurmBatchSystemTest::testCheckResourceRequest PASSED
src/toil/test/batchSystems/batchSystemTest.py::SlurmBatchSystemTest::testGetRescueJobFrequency PASSED
src/toil/test/batchSystems/batchSystemTest.py::SlurmBatchSystemTest::testRunJobs PASSED
src/toil/test/batchSystems/batchSystemTest.py::SlurmBatchSystemTest::testSetEnv PASSED
```

I suspect it's related to the slurm version. The slurm package available for Trusty 14.04 is 2.6 http://packages.ubuntu.com/trusty/slurm-llnl, but Vivid 15.04 has version 14.03 http://packages.ubuntu.com/vivid/slurm-wlm. In 2013, slurm switched version numbering schemes for yearly releases. I'll bring my instance back up and see what happens there.

**Comment by dleehr** [comment]
The issue was that the slurm config on the jenkins slave did not set the CPU and memory per node. The batchSystemTests do submit jobs with memory and CPU requirements, and slurm was not able to handle these requests. I've fixed the config in https://github.com/BD2KGenomics/cgcloud/pull/137 but we'll need to rebuild the jenkins slave image from that before this can be retried.

**Comment by hannes-ucsc** [comment]
I merged the cgcloud PR and updated the toil-jenkins-slave image. Now rerunning the tests.

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by ohofmann** [comment]
Following this with interest as well. We'd love to start using/testing Toil on a local SLURM system.

**Comment by hannes-ucsc** [comment]
src/toil/test/batchSystems/batchSystemTest.py::SlurmBatchSystemTest::testRunJobs seems to hang

**Comment by dleehr** [comment]
Yep, I see that. I need to check the subprocess for an exit code and fail if it failed. Then we should see why it's not working.

**Comment by hannes-ucsc** [comment]
Adding -s to the pytest command line disables output capturing

```diff
diff --git a/Makefile b/Makefile
index 8bc81e8..e0b0c2f 100644
--- a/Makefile
+++ b/Makefile
@@ -69,7 +67,7 @@ clean_sdist:
 
 
 test: check_venv
-       $(python) setup.py test --pytest-args "-vv $(tests)"
+       $(python) setup.py test --pytest-args "-vv -s $(tests)"
```

**Comment by dleehr** [comment]
Good tip, thanks

**Comment by brainstorm** [comment]
Sorry for the OT? But I'm getting Access Denied when registering an account on:

http://jenkins.cgcloud.info/user/brainstorm

And trying to access:

http://jenkins.cgcloud.info/job/toil-pull-requests/950/

Are those CI build logs public?

**Comment by dleehr** [comment]
That helped. It seems the jobs aren't starting. I need to make a couple more changes to the jenkins slave config, sorry!

**Comment by hannes-ucsc** [comment]
@brainstorm

> Are those CI build logs public?

No. Too risky. Only members of our GH org should have access. Is there a particular reason you would like to see them?

**Comment by ohofmann** [comment]
Mostly we are waiting for this implementation to build successfully (see https://github.com/guillermo-carrasco/toil/blob/slurm/src/toil/batchSystems/slurm.py and https://github.com/BD2KGenomics/toil/issues/682) and were wondering if there's some way for us to help with the debugging process.

**Comment by brainstorm** [comment]
@hannes-ucsc, I'm perhaps too used to CI/CD systems like [TravisCI](https://travis-ci.org/) and [CircleCI](https://circleci.com/) which have all logs available for every other open source project developer to see by default ;)

**Comment by dleehr** [comment]
@hannes-ucsc This will need another rebuild of the toil-jenkins-slave image, using https://github.com/BD2KGenomics/cgcloud/pull/141. I built a toil-jenkins-slave with that branch and the tests passed.

**Comment by hannes-ucsc** [comment]
@dleehr, I merged that PR and recreated the AMI.

Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
Jenkins, test this please.

**Comment by dleehr** [comment]
Thanks for merging/recreating. I don't know why this fails, when I build the image and run the tests myself, it passes. Perhaps someone with more slurm experience can take a look? /cc: @ohofmann @brainstorm  @guillermo-carrasco. I can cut out the appropriate parts of the log

**Comment by dleehr** [comment]
[slurm-toil-log.txt](https://github.com/BD2KGenomics/toil/files/213057/slurm-toil-log.txt)

**Comment by guillermo-carrasco** [comment]
@dleehr thanks for the log. However I can't find any useful information in there, could you please attach here, if possible, the actual SLURM log? 
 
Thanks!

**Comment by dleehr** [comment]
@guillermo-carrasco: I don't have access to the host or the slurm log. All I have is the console output from the test. @hannes-ucsc - Any ideas?

**Comment by guillermo-carrasco** [comment]
Just to make the point, we've been testing your branch in our cluster and it seems to work as well.

**Comment by brainstorm** [comment]
@dleehr I guess that a quick debug hack would be to run `slurmctld log` (if slurmd runs on the same test machine and if you are root).

**Comment by dleehr** [comment]
Thanks @brainstorm. I've now built a jenkins cluster with cgcloud (1 master and one toil slave). I'm seeing the same issues here, so I should be able to troubleshoot

**Comment by hannes-ucsc** [comment]
>  I've now built a jenkins cluster with cgcloud (1 master and one toil slave). I'm seeing the same issues here, so I should be able to troubleshoot

Glad to hear that you an reproduce this on your own cluster.

Generally speaking, cgcloud prefers to use a special application user as opposed to running applications as root or a user that has password-less sudo. Jenkins builds are run as the `jenkins` user which has sudo only for specific commands. You might have set up Slurm on the slave in such a way that only root can run certain commands. Ideally, an unprivileged user should be able to run Toil on a Slurm cluster, and the tests running as the unprivileged `jenkins` user would replicate that mode of operation. As a last resort, though, we could allow `jenkins` to run those privileged SLURM commands via `/etc/sudoers`. If you want to go down that route, search for `/etc/sudoers` in the cgcloud code base and you'll see what I mean.

**Comment by dleehr** [comment]
> Ideally, an unprivileged user should be able to run Toil on a Slurm cluster, and the tests running as the unprivileged `jenkins` user would replicate that mode of operation

Yep, and that's mostly working. The problem is with permissions on the `/var/run/slurm-llnl/slurm-acct.txt` file. The `sacct` command (used to check status of completed jobs) reads this file. If the user running sacct can't read the file, it fails. I added some debugging:

    DEBUG:toil.batchSystems.slurm:Checking status of internal job id 1
    DEBUG:toil.batchSystems.slurm:Getting exit code for slurm job 3
    DEBUG:toil.batchSystems.slurm:Checking line for exit codes: /var/run/slurm-llnl/slurm-acct.txt: Permission denied
    DEBUG:toil.batchSystems.slurm:Unable to get exit code

Despite setting the permission and ownership to `644` of this file in [toil_jenkins_slave.py](https://github.com/BD2KGenomics/cgcloud/blob/8bb4786f598d2bfefb2f8a486ae3bc0554e8ee03/jenkins/src/cgcloud/jenkins/toil_jenkins_slave.py#L314), permissions are `600` when jenkins runs the tests. Perhaps by virtue of being in [/var/run](http://askubuntu.com/questions/303120/how-folders-created-in-var-run-on-each-reboot). I didn't catch it when creating the slave from an image (using cgcloud) because cgcloud still runs the setup script every time (which fixes the perms)

I'll either relocate the file or otherwise ensure the permissions.

**Comment by hannes-ucsc** [comment]
> I'll either relocate the file or otherwise ensure the permissions.

Consider using `self._register_init_command()` for that. There is one usage of it already in  `JenkinSlave._setup_build_user()` which is similar to your use case.

Another alternative would be to modify the Slurm init scripts to do the necessary chmod/chown. 

Do the Slurm daemons need to be run as root?

**Comment by dleehr** [comment]
> Do the Slurm daemons need to be run as root?

`slurmctld` runs as slurm, and `slurmd` runs as root. It wasn't really a root problem, the file in question was owned by `slurm:slurm. That works fine, but the permissions were being set to 0600 on boot. I relocated the file to /var/log and it's happy there (https://github.com/BD2KGenomics/cgcloud/pull/145).

Thanks for your patience. I'm confident this will pass once the toil-jenkins-slave image is rebuilt again - I confirmed it on a cgcloud jenkins cluster.

**Comment by hannes-ucsc** [comment]
> I relocated the file to /var/log

Not exactly LSB, but ok. I may open a separate ticket to move it back to /var/run and fix the permissions explicitly.

**Comment by hannes-ucsc** [comment]
I created a new slave image with @dleehr's PRed changes.

Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
All tests pass! But build fails with 

```
You have are untracked files:
slurm-2.out
slurm-3.out
slurm-4.out
slurm-5.out
slurm-6.out
```

Toil tests have to clean up after themselves.

**Comment by hannes-ucsc** [comment]
And, looking at the history, I think the commits should be squashed.

**Comment by dleehr** [comment]
Ok I'll clean up the extra files. As for the commit squashing, GitHub can do that on merge now, does that work?

**Comment by hannes-ucsc** [comment]
> As for the commit squashing, GitHub can do that on merge now, does that work?

Doesn't that require you to rebase your branch on top of master? And if that is the case, how hard is it to squash while rebasing? Maybe I'm confused about this option. Another thing about squashing before the merge is that I get a CI build of the exact commit that get's merged. But again, educate me, I'm probably not up-to-date here.

**Comment by dleehr** [comment]
> Doesn't that require you to rebase your branch on top of master?

Actually no. I just tested it in a demo repo - It's a [new feature on GitHub](https://github.com/blog/2141-squash-your-commits). Seems to work as long as there are no conflicts.

I'll be pushing changes to clean up the dirty working copy very soon

File: src/toil/batchSystems/slurm.py [hunk 1]
```diff
@@ -212,17 +212,7 @@ def run(self):
...
-        # -b is binding y - not relevant for slurm
-        # -terse means just return the job id. Can I use -Q?
-        # -j means merge stderr + stdout (not necessary)
-        # -cwd means use the current working directory (not necessary)
-        # -o means standard output, but I think I can redirect this to /dev/null (done
-        # -e means standard error (done)
-        # -N is job name, which should be based on the jobID
-
-        sbatch_line = ['sbatch', '-Q', '-J', 'toil_job_{}'.format(jobID)]
+        sbatch_line = ['sbatch', '-Q', '-J', 'toil_job_{}'.format(jobID), '-o', '/dev/null', '-e', '/dev/null']
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/slurm.py]: This is what the GridEngine batch system does and turned out to be a bad idea. If anything goes wrong before worker.py gets a chance to setup its own redirection, any errors that occur will be swallowed, making diagnosing  problems much harder. 

We have something in the works that will allow us to set a workflow-specific location for files like this. For the time being, could you undo this change and instead add the appropriate clean-up code to the tests?

**Comment by hannes-ucsc** [comment]
> Seems to work as long as there are no conflicts.

I tried this too (on a different project) and it works. I don't like how it leaves the source branch dangling in the history. That and the fact that our CI does not build the exact merge commit will probably mean that we won't be using this feature for now. If you don't have the time to squash the commits, I will go ahead and do that in a separate PR (you'll remain the commit author). You have contributed a great deal and I don't want to test your patience with my pedantry.

**Comment by dleehr** [comment]
Ok I can squash when I update with the cleanup change. Happy to help.

**Comment by dleehr** [comment]
Looks like the only failure is `src.toil.test.cwl.cwlTest.CWLTest.test_run_conformance`, but that's failing in a bunch of other branches now too.

**Comment by hannes-ucsc** [comment]
Yes, test failure tracked in #783. Fix has been merged. Going to rebase this branch.

**Comment by hannes-ucsc** [comment]
Continued in #799

**Pull request closed without merging**